### PR TITLE
docs: Add scopes and auth ADR

### PR DESCRIPTION
**Pull request by oskarwilliams378**

### CODE REVIEW

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+Date: 2023-01-03
+
+## Status
+
+Draft
+
+## Context
+
+There are several routes we have explored for the moment in regard to Authentication and Authorisation.
+1. Have all auth exist in cognito and authorise with scopes
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:15]:
```suggestion
1. Have all auth exist in Cognito and authorise with scopes
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+2. Have all auth exist locally
+3. Have authentication exist in cognito, with very basic scopes, and then use our own database to narrow down subscription privileges.
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:17]:
```suggestion
3. Have authentication exist in Cognito, with very basic scopes, and then use our own database to narrow down subscription privileges
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Option 3 is what is being suggested for the moment, as this means we can use cognito to provide all the hard parts of auth, but use
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:19]:
```suggestion
Option 3 is what is being suggested for the moment, as this means we can use Cognito to provide all the hard parts of auth, but use
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+our database for enrichment and subscription privileges as this is easier to work with at high speed.
+
+## Current plan
+
+### Scopes in cognito
+
+1. events/publish - Allows the user to publish events to the system
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:26]:
```suggestion
1. `events/publish` - Allows the user to publish events to the system 
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+Option 3 is what is being suggested for the moment, as this means we can use cognito to provide all the hard parts of auth, but use
+our database for enrichment and subscription privileges as this is easier to work with at high speed.
+
+## Current plan
+
+### Scopes in cognito
+
+1. events/publish - Allows the user to publish events to the system 
+2. events/consume - Allows the user to consume events from the system
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:27]:
```suggestion
2. `events/consume` - Allows the user to consume events from the system
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+3. events/admin - Allows the user to use the admin and subscription features
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:28]:
```suggestion
3. `events/admin` - Allows the user to use the admin and subscription features
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+Date: 2023-01-03
+
+## Status
+
+Draft
+
+## Context
+
+There are several routes we have explored for the moment in regard to Authentication and Authorisation.
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:14]: I think I'd also explain that we currently do auth in a mix of places through Cognito and internally, and this is making it relatively tricky to manage

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+Each `Consumer` will have many `ConsumerSubscription`s, which will have an entry for `enrichmentFields` and an `eventType`.
+These `enrichmentFields` will decide what extra data past the event ID that the consumer will receive upon calling for events.
+
+Each `Publisher` will have many `PublisherSubscription`s, which will have an entry for `eventType` and a `datasetId`.
+The `eventType` will show which event will be published, and the `datasetId` shows where enrichment data will come from.
+
+### Limitations
+
+1. We will have to have extra DB tables
+2. We will have 2 sources of Auth
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:41]:
```suggestion
2. We are splitting some authorisation decisions between two places (Cognito and internal)
```

File: doc/architecture/decisions/0006-scopes-and-auth.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+1. We will be able to easily update enrichmentFields for a consumer
+2. We will be able to easily modify the data structure if needed based on changing requirements
+3. We will be able to still use cognito to lock down Authentication and base Authorisation
+
+## Future possible plans
+
+In the future, the current plan is to update this to remove `Consumer`s, `ConsumerSubscription`s, `Publisher`s, and 
+`PublisherSubscription`s from the DB by using Cognito users with custom attributes (claims) to determine the event type,
+enrichment fields, and event dataset. This will be easier when we have a more fixed data structure, and will remove the 
+limitations of the current plan.
```
Comment (olevett) [inline doc/architecture/decisions/0006-scopes-and-auth.md:55]: Though isn't without some limitations (e.g. how to structure claims, limits on custom attributes etc)

**Review by olevett — APPROVED** [review]
Couple of minor/petty thoughts, but otherwise looks good to me

File: doc/architecture/decisions/0006-scopes-and-auth.md (hunk 7 above)
Comment (oskarwilliams378) [reply doc/architecture/decisions/0006-scopes-and-auth.md:14]: This is what we still be doing though? The current functionality and the upcoming plan is the same?

**Review by olevett — APPROVED** [review]

**Pull request merged**